### PR TITLE
Remove overloads

### LINKED ISSUE #4694: remove overloads
### context

`useQuery` and friends have many overloads in TypeScript - different ways how the function can be invoked. As an example, let's look at `useQuery`. You can call it in three different ways:

```
useQuery(queryKey, queryFn, options)
useQuery(queryKey, options)
useQuery(options)
```

The `queryKey` is the only required property. The queryFn is optional because you might have defined it globally. If you use the syntax where you only pass one object in, the queryKey is also required.

Not only is this tough to maintain, type wise, it also requires a runtime check to see which type the first and the second parameter are to correctly create options. Internally, we only work with one options object.

Further, we started to add more overloads for features that are only doable with overloads. For example, if you pass `initialData`, the return type of `data` will _not_ contain `undefined`. Because `initialData` can also be a function, we need another 3 overloads for each possibility to call `useQuery`, resulting in 9 total overloads.

The `useQuery.ts` file has 140 lines of code - only 3 of which are actual JavaScript.

Other functions suffer from the same problem, and overloads are also not consistent. For example, `queryCache.findAll` has overloads, but `mutationCache.findAll` does not.

### proposal

- _remove_ all overloads and only allow all functions to be called with a single signature - one object

### detailed changes

```diff
- useQuery(key, fn, options)
+ useQuery({ queryKey, queryFn, ...options })
// ⬆️ queryKey must be required

- useInfiniteQuery(key, fn, options)
+ useInfiniteQuery({ queryKey, queryFn, ...options })
// ⬆️ queryKey must be required

- useMutation(fn, options)
+ useQuery({ mutationFn, ...options })

- useIsFetching(key, filters)
+ useIsFetching({ queryKey, ...filters })

- useIsMutating(key, filters)
+ useIsMutating({ mutationKey, ...filters })
```

```diff
- queryClient.isFetching(key, filters)
+ queryClient.isFetching({ queryKey, ...filters })

- queryClient.ensureQueryData(key, fn, options)
+ queryClient.ensureQueryData({ queryKey, queryFn, ...options })
// ⬆️ queryKey must be required

- queryClient.getQueriesData(key, filters)
+ queryClient.getQueriesData({ queryKey, ...filters })

- queryClient.setQueriesData(key, updater, filters, options)
+ queryClient.setQueriesData({ queryKey, ...filters }, updater, options)

- queryClient.removeQueries(key, filters)
+ queryClient.removeQueries({ queryKey, ...filters })

- queryClient.resetQueries(key, filters, options)
+ queryClient.resetQueries({ queryKey, ...filters }, options)

- queryClient.cancelQueries(key, filters, options)
+ queryClient.cancelQueries({ queryKey, ...filters }, options)

- queryClient.invalidateQueries(key, filters, options)
+ queryClient.invalidateQueries({ queryKey, ...filters }, options)

- queryClient.refetchQueries(key, filters, options)
+ queryClient.refetchQueries({ queryKey, ...filters }, options)

- queryClient.fetchQuery(key, fn, options)
+ queryClient.fetchQuery({ queryKey, queryFn, ...options })

- queryClient.prefetchQuery(key, fn, options)
+ queryClient.prefetchQuery({ queryKey, queryFn, ...options })

- queryClient.fetchInfiniteQuery(key, fn, options)
+ queryClient.fetchInfiniteQuery({ queryKey, queryFn, ...options })

- queryClient.prefetchInfiniteQuery(key, fn, options)
+ queryClient.prefetchInfiniteQuery({ queryKey, queryFn, ...options })

```

```diff
- queryCache.find(key, filters)
+ queryCache.find({ queryKey, ...filters })
// ⬆️ queryKey must be required
// `exact` could be omitted from possible filters and always set to `true` !

- queryCache.findAll(key, filters)
+ queryCache.findAll({ queryKey, ...filters })
```

### PR DESCRIPTION
**Pull request by Mamoanwar97**
This PR attempts to fix #4694 of v5 milestones, In this PR:

- remove overloads from query-core
- remove overloads from react-query
- remove overloads from solid-query
- remove overloads from vue-query
- update tests, docs, and examples accordingly

### CODE REVIEW

File: docs/react/adapters/solid-query.md [hunk 1]
```diff
@@ -69,10 +69,10 @@ Solid Query offers an API similar to  React Query, but there are some key differ
 
 ```tsx
 // ❌ react version
-useQuery(["todos", todo], fetchTodos)
+useQuery({queryKey: ["todos", todo], queryFn: fetchTodos})
 
 // ✅ solid version
-createQuery(() => ["todos", todo()], fetchTodos)
+createQuery({queryKey: () => ["todos", todo()], queryFn: fetchTodos})
```
Comment (TkDodo) [inline docs/react/adapters/solid-query.md:75]: I'm not sure if this is technically true - I think the whole object we pass into `createQuery` should be a function?

See the roadmap discussion with @ardeora here:
- https://github.com/TanStack/query/discussions/4252#discussioncomment-3779788

Also, for consistency, let's please use multi line whenever we use the object syntax:

```
useQuery({
  queryKey: ["todos", todo],
  queryFn: fetchTodos,
})
```

File: docs/react/adapters/solid-query.md [hunk 2]
```diff
@@ -161,7 +161,9 @@ const queryClient = new QueryClient()
 
 function Example() {
   const [enabled, setEnabled] = createSignal(false)
-  const query = createQuery(() => ['todos'], fetchTodos, {
+  const query = createQuery({ 
+    queryKey: () => ['todos'],
+    queryFn: fetchTodos,
     // ❌ passing a signal directly is not reactive
     // enabled: enabled(),
```
Comment (TkDodo) [inline docs/react/adapters/solid-query.md:168]: I think this workaround is also unnecessary without the overloads. @ardeora can you have a look please

File: docs/react/guides/migrating-to-react-query-5.md [hunk 3]
```diff
@@ -7,6 +7,23 @@ title: Migrating to TanStack Query v5
...
+```diff
+- useQuery(key, fn, options)
++ useQuery({ queryKey, queryFn, ...options })
+
+- useInfiniteQuery(key, fn, options)
++ useInfiniteQuery({ queryKey, queryFn, ...options })
+
+- useMutation(fn, options)
++ useMutation({ mutationFn, ...options })
+```
```
Comment (TkDodo) [inline docs/react/guides/migrating-to-react-query-5.md:25]: can we list all affected functions here please

File: examples/solid/basic-typescript/src/index.tsx [hunk 4]
```diff
@@ -97,13 +99,11 @@ const getPostById = async (id: number): Promise<Post> => {
...
-    () => getPostById(postId),
-    {
-      enabled: !!postId,
-    },
-  )
+  return createQuery({
+    queryKey: () => ['post', postId],
+    queryFn: () => getPostById(postId),
+    enabled: !!postId,
+  })
```
Comment (TkDodo) [inline examples/solid/basic-typescript/src/index.tsx:106]: also here, I think this should be:

```
createQuery(() => ({ queryKey, queryFn }))
```

File: packages/query-core/src/queryClient.ts [hunk 5]
```diff
@@ -113,84 +107,29 @@ export class QueryClient {
   }
 
   getQueryData<TQueryFnData = unknown>(
-    queryKey: QueryKey,
-    filters?: QueryFilters,
+    filters: WithRequired<QueryFilters, 'queryKey'>,
   ): TQueryFnData | undefined {
-    return this.queryCache.find<TQueryFnData>(queryKey, filters)?.state.data
+    return this.queryCache.find<TQueryFnData>(filters)?.state.data
   }
```
Comment (TkDodo) [inline packages/query-core/src/queryClient.ts:113]: I wasn't sure about `getQueryData`. I've only ever used it with a queryKey and no other filters. If the key is required, do the other filters even make sense 🤔. Maybe it would be better to have this as just a function that lets you pass in a `queryKey` only?

What do you think @ardeora @DamianOsipiuk ? On the other hand, one object is the consistent way 😅

File: packages/query-core/src/queryClient.ts [hunk 6]
```diff
@@ -202,41 +141,28 @@ export class QueryClient {
...
       return undefined
     }
 
-    const parsedOptions = parseQueryArgs(queryKey)
-    const defaultedOptions = this.defaultQueryOptions(parsedOptions)
+    const defaultedOptions = this.defaultQueryOptions({ queryKey })
     return this.queryCache
       .build(this, defaultedOptions)
-      .setData(data, { ...options, manual: true })
+      .setData(data, { ...options, manual: true }) as TQueryFnData
```
Comment (TkDodo) [inline packages/query-core/src/queryClient.ts:155]: why is this change needed?

File: packages/query-core/src/queryClient.ts [hunk 7]
```diff
@@ -608,16 +348,21 @@ export class QueryClient {
   }
 
   setQueryDefaults(
-    queryKey: QueryKey,
-    options: QueryObserverOptions<unknown, any, any, any>,
+    options: WithRequired<
+      QueryObserverOptions<unknown, any, any, any>,
+      'queryKey'
+    >,
```
Comment (TkDodo) [inline packages/query-core/src/queryClient.ts:354]: also here, not sure if these functions should be changed 🤔 . Would users still know that they'd be setting options for a specific key if we have:

```
queryClient.setQueryDefaults({
  queryKey: ['todos'],
  staleTime: 1000
})
```

?  Maybe it would be better to keep this as two arguments, and instead have the `queryKey` be _omitted_ from the options?

File: packages/react-query/src/types.ts [hunk 8]
```diff
@@ -28,19 +29,19 @@ export interface UseBaseQueryOptions<
   TQueryData = TQueryFnData,
   TQueryKey extends QueryKey = QueryKey,
 > extends ContextOptions,
-    QueryObserverOptions<TQueryFnData, TError, TData, TQueryData, TQueryKey> {}
+    WithRequired<
+      QueryObserverOptions<TQueryFnData, TError, TData, TQueryData, TQueryKey>,
+      'queryKey'
+    > {}
```
Comment (TkDodo) [inline packages/react-query/src/types.ts:35]: why don't we make the `queryKey` required directly inside `QueryOptions` ?

https://github.com/TanStack/query/blob/2b01ac0828b8dcabfa1a1747f44aaeb11e9c28ad/packages/query-core/src/types.ts#L76

Is there ever a situation where we want the `queryKey` to be optional there? I think it was only optional because we had overloads and the key could be a separate parameter ...

File: packages/solid-query/src/utils.ts [hunk 9]
```diff
@@ -1,61 +1,33 @@
...
-    return arg1 as any
-  }
-
-  if (typeof arg2 === 'function') {
-    return { ...arg3, queryKey: arg1(), queryFn: arg2 } as any
   }
-
-  return { ...arg2, queryKey: arg1() } as any
+  return arg as any
 }
```
Comment (TkDodo) [inline packages/solid-query/src/utils.ts:18]: not sure if this is still necessary. from what I can see, it tries to make the queryKey reactive ? I guess we'd now want that for the whole options object? Another question for @ardeora :)

File: packages/solid-query/src/createMutation.ts [hunk 10]
```diff
@@ -95,8 +42,8 @@ export function createMutation<
   })
 
   createComputed(() => {
-    const newParsedOptions = parseMutationArgs(arg1, arg2, arg3)
-    setOptions(newParsedOptions)
+    const newParsedOptions = { ...mutationOptions }
+    setOptions(mutationOptions)
```
Comment (TkDodo) [inline packages/solid-query/src/createMutation.ts:46]: is it necessary to create a shallow copy of `mutationOptions` ?

File: packages/vue-query/src/useBaseQuery.ts [hunk 11]
```diff
@@ -46,15 +45,14 @@ export function useBaseQuery<
...
-  arg3: UseQueryOptionsGeneric<TQueryFnData, TError, TData, TQueryKey> = {},
+  genericOptions: UseQueryOptionsGeneric<
+    TQueryFnData,
+    TError,
+    TData,
+    TQueryKey
+  >,
 ): UseQueryReturnType<TData, TError> {
-  const options = computed(() => parseQueryArgs(arg1, arg2, arg3))
+  const options = computed(() => parseQueryArgs(genericOptions))
```
Comment (TkDodo) [inline packages/vue-query/src/useBaseQuery.ts:55]: should `parseQueryArgs` be renamed now, as it doesn't do any parsing anymore? Maybe `unrefQueryArgs` ? @DamianOsipiuk

**Review by TkDodo — COMMENTED** [review]
awesome work, this is huge 🙌 . Please have a look at my comments

File: packages/react-query/src/types.ts (hunk 8 above)
Comment (Mamoanwar97) [reply packages/react-query/src/types.ts:35]: `QueryOptions` is used everywhere and extended by many things .. As a safe option, I made the `queryKey` only required in the places we need it.

Here are some examples where I think we don't need the queryKey to be required:

default options

HydrateOptions

getNextPageParam

fetch-In query-core/query.ts

getDefaultState in query-core/query.ts

File: packages/solid-query/src/createMutation.ts (hunk 10 above)
Comment (Mamoanwar97) [reply packages/solid-query/src/createMutation.ts:46]: I don't think so, I have just tried to remove it and the tests are passing .. will remove the shallow copy

File: packages/query-core/src/queryClient.ts (hunk 6 above)
Comment (Mamoanwar97) [reply packages/query-core/src/queryClient.ts:155]: for a reason, without the `as TQueryFnData`, typescript reads that `setData` is returning `unknown`

without-as-TQueryFnData

While inside `setData` it infers the first argument type and returns it so it should return the `TQueryFnData` 

setData

but for a reason that is not happening when we use setData inside setQueryData, although the type of `data` passed to the `setData` is `TQueryFnData` not `unknown`

typof data passed to setData

File: packages/vue-query/src/useBaseQuery.ts (hunk 11 above)
Comment (Mamoanwar97) [reply packages/vue-query/src/useBaseQuery.ts]: Done ✅

File: packages/solid-query/src/createMutation.ts (hunk 10 above)
Comment (Mamoanwar97) [reply packages/solid-query/src/createMutation.ts]: Done ✅

File: packages/query-core/src/queryClient.ts (hunk 7 above)
Comment (Mamoanwar97) [reply packages/query-core/src/queryClient.ts]: Done ✅

File: docs/react/adapters/solid-query.md (hunk 1 above)
Comment (Mamoanwar97) [reply docs/react/adapters/solid-query.md]: Done ✅

File: docs/react/guides/migrating-to-react-query-5.md (hunk 3 above)
Comment (Mamoanwar97) [reply docs/react/guides/migrating-to-react-query-5.md]: Done ✅

File: examples/solid/basic-typescript/src/index.tsx (hunk 4 above)
Comment (Mamoanwar97) [reply examples/solid/basic-typescript/src/index.tsx:106]: It is my first time dealing with solid tbh, but I think the object syntax is working fine with the tests 

@ardeora is the object syntax fine? or should it be converted to an arrow function returning the object?

File: packages/query-core/src/queryClient.ts (hunk 6 above)
Comment (TkDodo) [reply packages/query-core/src/queryClient.ts:155]: ah yes, so would:

```
.setData<TQueryFnData>(data, { ...options, manual: true })
```

work as well?

File: packages/react-query/src/types.ts (hunk 8 above)
Comment (TkDodo) [reply packages/react-query/src/types.ts:35]: that's fine, we can revisit this later if necessary. There are some functions that don't really need the full `QueryOptions` (like `getNextPageParam`).

File: packages/query-core/src/queryClient.ts (hunk 6 above)
Comment (Mamoanwar97) [reply packages/query-core/src/queryClient.ts]: Fixed ✅ 

I traced it and tried to find the difference between my branch and the main branch 

In the main branch:
- queryKey was passed to parseQueryArgs which returns a type of `QueryOptions<any, any, any, QueryKey>`
- when the `parsedOptions` is passed to `defaultQueryOptions` it returns ` DefaultedQueryObserverOptions<any, any, unknown, any, QueryKey>`

main-branch

but in my branch, because I was passing the `queryKey` only to `defaultQueryOptions` it inferred the type to be `DefaultedQueryObserverOptions<unknown, unknown, unknown, unknown, QueryKey>`

my-branch-before

so to fix it instead of `as` alias, I passed the generics to the `defaultQueryOptions` by myself

my-branch after

File: packages/query-core/src/queryClient.ts (hunk 5 above)
Comment (DamianOsipiuk) [reply packages/query-core/src/queryClient.ts:113]: I would opt for either:
- single argument with everything `optional`. As an alias to `getQueryCache().find()`
- single argument of `queryKey`. Then if you need more advanced search you would do `getQueryCache().find()`.

To be honest i would go with only `queryKey` as if i would like to use more filters i would expect to potentially get more than one result. Meaning `getQueriesData`.

File: packages/vue-query/src/useBaseQuery.ts (hunk 11 above)
Comment (DamianOsipiuk) [reply packages/vue-query/src/useBaseQuery.ts]: I would use `cloneDeepUnref` directly here and in other places.
Also with some minor changes to types of `cloneDeepUnref` and `cloneDeep` we could get rid of many `as` assertions all over the codebase.
But i guess i could do that in another PR.
This should be good enough. 👍

File: examples/solid/basic-typescript/src/index.tsx (hunk 4 above)
Comment (ardeora) [reply examples/solid/basic-typescript/src/index.tsx:106]: Thanks for working on this! Looks great. Yeah solid query `v5` will need more changes here. With the removal of overloads. We don't need to have `QueryKey` as a function anymore and would be more like what Dominik mentioned above. For example,

```ts
createQuery({
    queryKey: () => ['post', postId],
    queryFn: () => getPostById(postId),
    enabled: !!postId,
  })
```

will become 

```ts
createQuery(() => {
    queryKey: ['post', postId],
    queryFn: () => getPostById(postId),
    enabled: !!postId,
  })
```

But this change is definitely wayyy more involved since we need to change the typings in `types.ts` to get in line with with `react-query`. It would also change how `createQuery` tracks the signal changes inside the function. This would also mean changing the tests to reflect the same. @TkDodo  do you suggest we make those changes in a separate PR or should I add annotations on this PR itself?

File: packages/query-core/src/queryClient.ts (hunk 5 above)
Comment (ardeora) [reply packages/query-core/src/queryClient.ts:113]: Yeah I agree with @DamianOsipiuk. For example, if I use the filters like `exact: false` or `predicate`, I'm usually expecting more than one result

File: examples/solid/basic-typescript/src/index.tsx (hunk 4 above)
Comment (TkDodo) [reply examples/solid/basic-typescript/src/index.tsx:106]: I see 3 ways forward here:

1) you can checkout the PR with the gh cli and work on it directly
2) you can create a PR _to_ this PR :)
3) we can merge this to v5, as it will not be released, and you can work on it in a separate issue / PR

whatever you prefer is fine with me :)

File: packages/query-core/src/queryClient.ts (hunk 5 above)
Comment (TkDodo) [reply packages/query-core/src/queryClient.ts:113]: agree. It's weird that we would only get the first matching result then - that's what we have `getQueriesData` etc. for.

Ok then let's do one param only, that is _just_ the queryKey, and that is required then.

Comment (Mamoanwar97) [reply packages/query-core/src/queryClient.ts:113]: sure 🙌 should I apply the same concept on `getQueryState` ? I think it is similar to `getQueryData`

Comment (TkDodo) [reply packages/query-core/src/queryClient.ts]: yes, they are pretty much the same :)

Comment (Mamoanwar97) [reply packages/query-core/src/queryClient.ts]: Done ✅ also updated the migration docs to include this change 🙌

**Review by DamianOsipiuk — APPROVED** [review]

File: examples/solid/basic-typescript/src/index.tsx (hunk 4 above)
Comment (DamianOsipiuk) [reply examples/solid/basic-typescript/src/index.tsx:104]: I vote for merging it if it's good enough for now.
This PR is massive and it would be hard to keep up to date, when other PRs start merging to v5.

Comment (TkDodo) [reply examples/solid/basic-typescript/src/index.tsx:104]: yes, you're right @DamianOsipiuk . There are already conflicts, sorry for that. Once they are fixed, I'll merge this PR next

Comment (ardeora) [reply examples/solid/basic-typescript/src/index.tsx:104]: Yeah, I agree! The scope for the other `solid-query` v5 changes can be addressed in a separate PR

**Review by ardeora — APPROVED** [review]

File: examples/solid/basic-typescript/src/index.tsx (hunk 4 above)
Comment (Mamoanwar97) [reply examples/solid/basic-typescript/src/index.tsx:104]: the merge conflicts are fixed now 🫡

Comment (Mamoanwar97) [reply examples/solid/basic-typescript/src/index.tsx:104]: @TkDodo merge conflicts are fixed 🫡

**Comment by TkDodo** [comment]
@allcontributors add @Mamoanwar97 for code

**Comment by TkDodo** [comment]
@ardeora merged. Please create a follow up PR to address the open solid todos (or just an issue if you'd want someone else to tackle it). @DamianOsipiuk I don't think there is anything open for vue, but if there is, let's also create a follow-up
